service/src/LevelZeroDevicePool.cpp

```diff
@@ -106,6 +106,25 @@ namespace geopm
                                             dev_subdev_idx_pair.second);
     }
 
+    double LevelZeroDevicePoolImp::frequency_efficient(int domain, unsigned int domain_idx,
+                                                       int l0_domain) const
+    {
+        if (domain != GEOPM_DOMAIN_GPU_CHIP) {
+            throw Exception("LevelZeroDevicePool::" + std::string(__func__) +
+                            ": domain " + std::to_string(domain) +
+                            " is not supported for the frequency domain.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        std::pair<unsigned int, unsigned int> dev_subdev_idx_pair;
+        dev_subdev_idx_pair = subdevice_device_conversion(domain_idx);
```

Comment:
How does a "domain_idx" map to a subdevice-device pair? How are these pairings being tracked?

Reply:
Each GPU has a number of GPU_CHIPs.  This mapping is tracked in the LevelZero..cpp device struct provided via the num_gpu function.  
The subdevice_device_conversion function does a simple split of number of sub-devices to devices.